### PR TITLE
[RFC] Support optimizing over any Tensor with requires_grad = True

### LINKED ISSUE #1604: Simple training script with loss cause an exception
The following steps causes a runtime exception `NotImplementedError`. But, I am sorry in advance if I make mistakes. Here is a repro. 
`"eagar"` or `"aot_eagar`" instead of "`inductor`" does not work, too.

```
% $ git log | head -1
commit 8c9f11ca6f2789f06785e7606bdb99f087bcc73a
% pip list | grep torch
torch              1.13.0.dev20220927+cpu
torch-mlir         20220927.609
torchdynamo        1.13.0.dev0            /mnt/xvdc/DeepTools/DD2/torchdynamo
torchvision        0.14.0.dev20220927+cpu
% cat torchdynamo_loss.py
import torch
import torchdynamo

class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y):
        x = torch.bmm(x, y)
        x = torch.flatten(x, 1)
        return x

@torchdynamo.optimize("inductor")
def training_iter_fn(batch, model, optimizer):
    optimizer.zero_grad()
    out = model(**batch)
    lossFn = torch.nn.CrossEntropyLoss()
    target = torch.tensor([0, 7])
    loss = lossFn(out, target)
    loss.backward()
    optimizer.step()
    return loss

net = Net()
input1 = torch.randn(2, 1, 4)
input2 = torch.randn(2, 4, 8, requires_grad=True)
optimizer = torch.optim.Adam([input2], lr=0.1)

opt_training_iter_fn = training_iter_fn
batch = {"x":input1, "y":input2}
loss = opt_training_iter_fn(batch, net, optimizer)

print(input2.cpu())


% python torchdynamo_loss.py 
[2022-10-12 06:42:13,178] torchdynamo.variables.torch: [WARNING] Profiler will be ignored
[2022-10-12 06:42:13,187] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.backward from user code at   File "/home/ishizaki/torchdynamo/tmp/torchdynamo_loss.py", line 26, in training_iter_fn
    loss.backward()

[2022-10-12 06:42:14,922] torchdynamo.symbolic_convert: [WARNING] Graph break: inline with __closure__ from user code at   File "/home/ishizaki/torchdynamo/tmp/torchdynamo_loss.py", line 27, in <graph break in training_iter_fn>
    optimizer.step()

[2022-10-12 06:42:14,934] torchdynamo.symbolic_convert: [WARNING] Graph break: inline in skipfiles: _fn /home/ishizaki/torchdynamo/torchdynamo/eval_frame.py from user code at   File "/home/ishizaki/torchdynamo/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 178, in step
    self._cuda_graph_capture_health_check()

[2022-10-12 06:42:14,948] torchdynamo.convert_frame: [ERROR] WON'T CONVERT <graph break in step> /home/ishizaki/torchdynamo/.venv/lib/python3.9/site-packages/torch/optim/adam.py line 178 
due to: 
Traceback (most recent call last):
  File "/home/ishizaki/torchdynamo/torchdynamo/variables/base.py", line 146, in as_python_constant
    raise NotImplementedError(f"{self} is not a constant")
NotImplementedError: TensorVariable() is not a constant

from user code:
   File "/home/ishizaki/torchdynamo/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 209, in <graph break in step>
    state = self.state[p]

Set torchdynamo.config.verbose=True for more information
==========
[2022-10-12 06:42:14,961] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.item from user code at   File "/home/ishizaki/torchdynamo/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 300, in adam
    func(params,
  File "/home/ishizaki/torchdynamo/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 395, in _single_tensor_adam
    step = step_t.item()

[2022-10-12 06:42:14,980] torchdynamo.optimizations.training: [WARNING] Unable to use Aot Autograd because of presence of mutation
[2022-10-12 06:42:14,980] torchinductor.compile_fx: [WARNING] Aot Autograd is not safe to run, so falling back to eager
[2022-10-12 06:42:15,000] torchdynamo.optimizations.training: [WARNING] Unable to use Aot Autograd because of presence of mutation
[2022-10-12 06:42:15,000] torchinductor.compile_fx: [WARNING] Aot Autograd is not safe to run, so falling back to eager
tensor([[[ 1.2348, -0.7126,  1.2387,  2.0989, -0.1772,  0.4236, -0.0968,
          -0.3471],
         [ 0.6295, -1.5259,  0.6826,  1.0028, -0.4873,  0.0893, -0.2904,
           0.1033],
         [ 1.7367,  0.7345,  1.5238, -1.9054, -1.9447,  0.4717,  0.1325,
          -0.6108],
         [ 0.5933,  0.7319,  1.5816, -0.3573,  0.3974, -1.0648, -2.0550,
           0.6247]],

        [[ 0.4393,  0.4159, -0.4996,  0.3288, -0.9796, -0.0822, -0.6735,
           0.4048],
         [-1.1754, -0.2157,  1.0433, -0.3781,  0.5304, -2.7421, -1.1731,
          -0.6624],
         [ 0.3439, -0.4731,  0.4820, -0.1286, -0.1511,  0.4843,  1.1936,
           1.2146],
         [ 1.9118, -1.4318, -0.6035,  0.0142,  0.8406,  1.2690, -0.2417,
           0.4326]]], requires_grad=True)
```

### PR DESCRIPTION
**Pull request by yanboliang**
Fix #1604 

I had discussed with @mlazos, the solution may has risk if tensor is not immutable. So we may need to add more checks before deciding specialize value. Open this PR for feedback.

### CODE REVIEW

File: test/dynamo/test_optimizers.py [hunk 1]
```diff
@@ -95,3 +95,37 @@ def setUpClass(cls):
...
+        net = Net()
+        input1 = torch.randn(2, 1, 4)
+        input2 = torch.randn(2, 4, 8, requires_grad=True)
+        optimizer = torch.optim.Adam([input2], lr=0.1)
+
+        cnts = torchdynamo.testing.CompileCounter()
+        opt_training_iter_fn = torchdynamo.optimize(cnts)(training_iter_fn)
+        batch = {"x": input1, "y": input2}
+        opt_training_iter_fn(batch, net, optimizer)
+        self.assertEqual(cnts.frame_count, 4)
```
Comment (voznesenskym) [inline test/dynamo/test_optimizers.py:131]: Why is this 4 frames? Where are the breaks?

**Comment by mlazos** [comment]
I chatted with Jason offline a bit. Basically I think we should narrow this to specialize when we are getting the tensors from the optimizer, and in other cases don't allow this. Right now from your PR it looks like you always specialize on them.

File: test/dynamo/test_optimizers.py (hunk 1 above)
Comment (yanboliang) [reply test/dynamo/test_optimizers.py:131]: Yea, I think these breaks are known issues:
```
[2022-10-14 05:19:22,296] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.backward from user code at   File "/scratch/ybliang/work/repos/torchdynamo/debug/debug1.py", line 53, in training_iter_fn
    loss.backward()

[2022-10-14 05:19:22,381] torchdynamo.symbolic_convert: [WARNING] Graph break: inline with __closure__ from user code at   File "/scratch/ybliang/work/repos/torchdynamo/debug/debug1.py", line 54, in <graph break in training_iter_fn>
    optimizer.step()

[2022-10-14 05:19:22,392] torchdynamo.symbolic_convert: [WARNING] Graph break: inline in skipfiles: _fn /scratch/ybliang/work/repos/torchdynamo/torchdynamo/eval_frame.py from user code at   File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 178, in step
    self._cuda_graph_capture_health_check()

[2022-10-14 05:19:22,428] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.item from user code at   File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 234, in <graph break in step>
    adam(params_with_grad,
  File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 300, in adam
    func(params,
  File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 395, in _single_tensor_adam
    step = step_t.item()
```
Some of the breaks, e.g, ```Tensor.item``` actually we can avoid by setting ```capturable=True```, but here I use the basic API for unit test.

**Comment by yanboliang** [comment]
> I chatted with Jason offline a bit. Basically I think we should narrow this to specialize when we are getting the tensors from the optimizer, and in other cases don't allow this. Right now from your PR it looks like you always specialize on them.

That makes sense, I agree we should guard on specific cases. We can pass in where the tensor comes from when creating and decide if specialize. I'll update this PR to reflect this change. Thank you!

**Comment by mlazos** [comment]
> > I chatted with Jason offline a bit. Basically I think we should narrow this to specialize when we are getting the tensors from the optimizer, and in other cases don't allow this. Right now from your PR it looks like you always specialize on them.
> 
> That makes sense, I agree we should guard on specific cases. We can pass in where the tensor comes from when creating and decide if to specialize. I'll update this PR to reflect this change. Thank you!

One way to implement it for you to consider - Create an OptimizerParameterVariable for the case where we read params from the optimizer object (I think it's safe to say that any optimizer would always inherit from `Optimizer`) and then in this case set `parameter_value` on the resulting tensor variable.

**Comment by jansel** [comment]
We have migrated `torchdynamo` to `torch._dynamo` and will use the `pytorch/pytorch` repo for future development.  Please resubmit this PR to https://github.com/pytorch/pytorch/

More details and instructions to port this PR over can be found in https://github.com/pytorch/torchdynamo/issues/1588

**Pull request closed without merging**